From: OpenMandriva
Subject: [PATCH] pyo3: Disable abi3 feature

OpenMandriva does not require compatibility with Python versions
different to the one at compile time. PyO3 abi3-py310 cannot load
into Python 3.15.

---
--- a/build_rust.py	2026-10-14 16:22:00.000000000 +0200
+++ b/build_rust.py	2026-10-16 19:16:32.410643476 +0200
@@ -1,10 +1,8 @@
 # A build script for poetry that adds the rust extension.
 
-import itertools
 import os
 from typing import Any
 
-from packaging.specifiers import SpecifierSet
 from setuptools_rust import Binding, RustExtension
 
 
@@ -16,27 +14,11 @@
         target="synapse.synapse_rust",
         path=cargo_toml_path,
         binding=Binding.PyO3,
-        # This flag is a no-op in the latest versions. Instead, we need to
-        # specify this in the `bdist_wheel` config below.
-        py_limited_api=True,
+        # This flag is a no-op in the latest versions.
+        py_limited_api=False,
         # We always build in release mode, as we can't distinguish
         # between using `poetry` in development vs production.
         debug=False,
     )
     setup_kwargs.setdefault("rust_extensions", []).append(extension)
     setup_kwargs["zip_safe"] = False
-
-    # We look up the minimum supported Python version with
-    # `python_requires` (e.g. ">=3.10.0,<4.0.0") and finding the first Python
-    # version that matches. We then convert that into the `py_limited_api` form,
-    # e.g. cp310 for Python 3.10.
-    py_limited_api: str
-    python_bounds = SpecifierSet(setup_kwargs["python_requires"])
-    for minor_version in itertools.count(start=10):
-        if f"3.{minor_version}.0" in python_bounds:
-            py_limited_api = f"cp3{minor_version}"
-            break
-
-    setup_kwargs.setdefault("options", {}).setdefault("bdist_wheel", {})[
-        "py_limited_api"
-    ] = py_limited_api
--- a/rust/Cargo.toml	2026-10-14 16:22:00.000000000 +0200
+++ b/rust/Cargo.toml	2026-10-16 19:16:32.410707497 +0200
@@ -35,8 +35,6 @@
 pyo3 = { version = "0.29.0", features = [
     "macros",
     "anyhow",
-    "abi3",
-    "abi3-py310",
     # So we can pass `bytes::Bytes` directly back to Python efficiently,
     # https://docs.rs/pyo3/latest/pyo3/bytes/index.html
     "bytes",
